framework/src/Volo.Abp.Emailing/Volo/Abp/Emailing/EmailSenderBase.cs

```diff
@@ -137,52 +87,139 @@ public virtual async Task QueueAsync(string from, string to, string subject, str
+                                        Encoding.UTF8);
         }
 
-        if (mail.HeadersEncoding == null)
-        {
-            mail.HeadersEncoding = Encoding.UTF8;
-        }
+        mail.HeadersEncoding ??= Encoding.UTF8;
 
-        if (mail.SubjectEncoding == null)
-        {
-            mail.SubjectEncoding = Encoding.UTF8;
-        }
+        mail.SubjectEncoding ??= Encoding.UTF8;
+
+        mail.BodyEncoding ??= Encoding.UTF8;
+    }
+
+    /// <summary>
+    /// Should implement this method to send email in derived classes.
+    /// </summary>
+    /// <param name="mail">Mail to be sent</param>
+    protected abstract Task SendEmailAsync(MailMessage mail);
 
-        if (mail.BodyEncoding == null)
+    private static void NormalizeMailForBase64Image(MailMessage mail)
+    {
+        var htmlBody = mail.Body;
+
+        // Find all base64 image tags in the html message.
+        var base64ImageTags = Regex.Matches(htmlBody, @"<img src=""data:image/(png|jpeg|gif);base64,([^""]+)""[^>]*>");
+
+        foreach (Match base64ImageTag in base64ImageTags)
         {
-            mail.BodyEncoding = Encoding.UTF8;
+            // Extract the base64-encoded image data from the tag.
+            var format = base64ImageTag.Groups[1].Value;
+            var base64Data = base64ImageTag.Groups[2].Value;
+
+            // Convert the base64 data to binary.
+            var imageByte = Convert.FromBase64String(base64Data);
+
+            // Create a unique image name and a MemoryStream containing the image data.
+            var imageName = Guid.NewGuid().ToString();
+            MemoryStream imageStream = new(imageByte);
+
+            // Replace the base64 image tag with cid: image tag.
+            var newImageTag = base64ImageTag.Value.Replace($"<img src=\"data:image/{format};base64,{base64Data}\"", $"<img src =\"cid:{imageName}\"");
+            // Create the HTML view
+            htmlBody = htmlBody.Replace(base64ImageTag.Value, newImageTag);
+            var htmlView = AlternateView.CreateAlternateViewFromString(
+                                                         htmlBody,
+                                                         Encoding.UTF8,
+                                                         MediaTypeNames.Text.Html);
+            // Create a plain text message for client that don't support HTML
+            var plainView = AlternateView.CreateAlternateViewFromString(htmlBody,
+                                                        Encoding.UTF8,
+                                                        MediaTypeNames.Text.Plain);
+
+            var img = new LinkedResource(imageStream, $"image/{format}")
+            {
+                ContentId = imageName
+            };
+            img.ContentType.MediaType = $"image/{format}";
+            img.TransferEncoding = TransferEncoding.Base64;
+            img.ContentType.Name = img.ContentId;
+            img.ContentLink = new Uri("cid:" + img.ContentId);
```

Comment:
Is `cid` a universal HTML format? or only Gmail understand this tag?

Reply:
No, "cid" is not a universal HTML format. CID stands for "Content-ID" and is used specifically in the context of email and HTML email messages.

In email, the "cid" is used to reference embedded content, such as images or other media, that is included within the email message itself. The CID acts as a unique identifier for that embedded content.

For example, in an HTML email, an image might be referenced using a CID like this:
`<img src="cid:image001.jpg">`

This tells the email client to look for an image with the content ID "image001.jpg" that is embedded within the email message.

However, this "cid" reference is not a standard HTML format that can be used outside of the email context. In regular web pages, images and other resources are typically referenced using standard URL paths, not CIDs.

So in summary, "cid" is specific to email and not a universal HTML format for referencing content on the web. Regular HTML uses other methods like \[src="image.jpg"\] to link to external resources.